test/e2e/user_group_limit/user_group_limit_test.go

```diff
@@ -909,40 +926,227 @@ var _ = ginkgo.Describe("UserGroupLimit", func() {
+				},
+			},
+			Type: v1.SecretTypeServiceAccountToken,
+		}
+		_, err = kClient.CreateSecret(secret, namespace)
+		gomega.Ω(err).NotTo(HaveOccurred())
+		// Get the token value from the Secret
+		ginkgo.By("Getting the token value from the Secret...")
+		userTokenValue, err := kClient.GetSecretValue(namespace, secretName, "token")
+		gomega.Ω(err).NotTo(HaveOccurred())
+		// use deep copy not to hardcode the kubeconfig
+		config, err := kClient.GetKubeConfig()
+		gomega.Ω(err).NotTo(HaveOccurred())
+		config.BearerToken = userTokenValue
+		newConf := rest.CopyConfig(config) // copy existing config
+		// Use token-based authentication instead of client certificates
+		newConf.CAFile = ""
+		newConf.CertFile = ""
+		newConf.KeyFile = ""
+		newConf.BearerToken = userTokenValue
+		kubeconfigPath := filepath.Join(os.TempDir(), "test-user-config")
+		err = k8s.WriteConfigToFile(newConf, kubeconfigPath)
+		gomega.Ω(err).NotTo(HaveOccurred())
+		config, err = clientcmd.BuildConfigFromFlags("", kubeconfigPath)
+		gomega.Ω(err).NotTo(HaveOccurred())
+		clientset, err = kubernetes.NewForConfig(config)
+		gomega.Ω(err).NotTo(HaveOccurred())
+		ginkgo.By("Creating Pod...")
+		pod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: podName,
+				Annotations: map[string]string{
+					"created-by": fmt.Sprintf("system:serviceaccount:%s:%s", namespace, serviceAccountName),
+					"user-token": userTokenValue, // Log the token in the annotation
+				},
+				Labels: map[string]string{"applicationId": "test-app"},
+			},
+			Spec: v1.PodSpec{
+				ServiceAccountName: serviceAccountName,
+				Containers: []v1.Container{
+					{
+						Name:  "nginx",
+						Image: "nginx",
+						Ports: []v1.ContainerPort{{ContainerPort: 80}},
+					},
+				},
+			},
+		}
+		_, err = clientset.CoreV1().Pods(namespace).Create(context.TODO(), pod, metav1.CreateOptions{})
+		gomega.Ω(err).NotTo(HaveOccurred())
+		createdPod, err := clientset.CoreV1().Pods(namespace).Get(context.TODO(), podName, metav1.GetOptions{})
+		gomega.Ω(err).NotTo(HaveOccurred())
+		ginkgo.By("Verifying User Info...")
+		userInfo, err := GetUserInfoFromPodAnnotation(createdPod)
+		gomega.Ω(err).NotTo(HaveOccurred())
+		// user info should contain the substring "system:serviceaccount:default:test-user-sa"
+		gomega.Ω(strings.Contains(fmt.Sprintf("%v", userInfo), "system:serviceaccount:default:test-user-sa")).To(gomega.BeTrue())
+		// Cleanup
+		ginkgo.By("Cleaning up resources...")
+		err = clientset.CoreV1().Pods(namespace).Delete(context.TODO(), podName, metav1.DeleteOptions{})
+		gomega.Ω(err).NotTo(HaveOccurred())
+		err = clientset.CoreV1().ServiceAccounts(namespace).Delete(context.TODO(), serviceAccountName, metav1.DeleteOptions{})
+		gomega.Ω(err).NotTo(HaveOccurred())
+		err = kClient.DeleteClusterRole("pod-creator-role")
+		gomega.Ω(err).NotTo(HaveOccurred())
+		err = kClient.DeleteClusterRoleBindings("pod-creator-role-binding")
+		gomega.Ω(err).NotTo(HaveOccurred())
```

Comment:
If the test fail, we never reach this part. This is supposed to be in a deferred section:

That is:
```
		defer func() {
				// Cleanup
				ginkgo.By("Cleaning up resources...")
				err = clientset.CoreV1().Pods(namespace).Delete(context.TODO(), podName, metav1.DeleteOptions{})
				gomega.Ω(err).NotTo(HaveOccurred())
				err = clientset.CoreV1().ServiceAccounts(namespace).Delete(context.TODO(), serviceAccountName, metav1.DeleteOptions{})
				gomega.Ω(err).NotTo(HaveOccurred())
				err = kClient.DeleteClusterRole("pod-creator-role")
				gomega.Ω(err).NotTo(HaveOccurred())
				err = kClient.DeleteClusterRoleBindings("pod-creator-role-binding")
				gomega.Ω(err).NotTo(HaveOccurred())
		}
		// Create Service Account
		ginkgo.By("Creating Service Account...")
		sa, err := kClient.CreateServiceAccount(serviceAccountName, namespace)
		gomega.Ω(err).NotTo(HaveOccurred())
```